### PR TITLE
Update the ResponseEndBlock values

### PR DESCRIPTION
**Pull request by ethanfrey**
* Rename diffs -> changes
* Add consensus_param_changes

For context, see:
https://github.com/tendermint/tendermint/issues/924
https://github.com/tendermint/tendermint/issues/952

### CODE REVIEW

**Comment by ethanfrey** [comment]
This is needed for other prs in tendermint/tendermint

File: README.md [hunk 1]
```diff
@@ -185,7 +185,8 @@ Here, we describe the requests and responses as function arguments and return va
   * __Arguments__:
     * `Height (int64)`: The block height that ended
   * __Returns__:
-    * `Diffs ([]Validator)`: Changed validators with new voting powers (0 to remove)
+    * `Changes ([]Validator)`: Changed validators with new voting powers (0 to remove)
+    * `ConsensusParamChanges (ConsensusParams)`: Communicate changes to consensus-critical time/size parameters to tendermint core
```
Comment (melekes) [inline README.md]: I'd remove `Communicate`. Just `Changes to ...`

Comment (jaekwon) [reply README.md]: If we're going to rename this, lets get more specific.  Like ValidatorUpdates or something.

Comment (jaekwon) [reply README.md]: s/Changes/Updates/g, more canonical.

File: types/types.proto [hunk 2]
```diff
@@ -157,7 +157,35 @@ message ResponseBeginBlock{
 }
 
 message ResponseEndBlock{
-  repeated Validator diffs = 1;
+  repeated Validator changes = 1;
+  ConsensusParams consensus_param_changes = 2;
+}
+
+message ConsensusParams{
+    BlockSizeParams block_size_params = 1;
```
Comment (jaekwon) [inline types/types.proto]: Propose s/BlockSizeParams/BlockSize/g etc.
ConsensusParams.BlockSize.MaxBytes, not
ConsensusParams.BlockSizeParams.MaxBytes.

Otherwise, what about nested params?
ConsensusParams.BlockSizeParams.SomethingElseParams.Thing?? We don't want that.

**Comment by ethanfrey** [comment]
This is breaking... can I merge it in, or should I wait for the downstream tendermint pr first? https://github.com/tendermint/tendermint/pull/972

File: README.md [hunk 3]
```diff
@@ -185,7 +185,8 @@ Here, we describe the requests and responses as function arguments and return va
   * __Arguments__:
     * `Height (int64)`: The block height that ended
   * __Returns__:
-    * `Diffs ([]Validator)`: Changed validators with new voting powers (0 to remove)
+    * `Updates ([]Validator)`: Updated validators with new voting powers (0 to remove)
```
Comment (melekes) [inline README.md]: Not `ValidatorUpdates`?

File: README.md [hunk 4]
```diff
@@ -185,7 +185,8 @@ Here, we describe the requests and responses as function arguments and return va
...
+    * `ConsensusParamChanges (ConsensusParams)`: Changes to consensus-critical time/size parameters to tendermint core
```
Comment (melekes) [inline README.md]: `ConsensusParamUpdates` to be consistent, no?

File: README.md (hunk 3 above)
Comment (melekes) [reply README.md]: Note: This is from Jae

**Comment by jaekwon** [comment]
LGTM.  Made some updates.

File: Makefile [hunk 5]
```diff
@@ -8,7 +8,7 @@ GOTOOLS = \
 
 INCLUDE = -I=. -I=${GOPATH}/src -I=${GOPATH}/src/github.com/gogo/protobuf/protobuf
 
-all: install test
+all: protoc install test metalinter
```
Comment (melekes) [inline Makefile]: `metalinter` - no such target! rename to `metalinter_test`

File: Makefile [hunk 6]
```diff
@@ -65,34 +63,34 @@ get_deps:
...
+	@ echo "--> Running glide install"
 	@ glide install
 
-metalinter:
+metalinter_all:
 	protoc $(INCLUDE) --lint_out=. types/*.proto
 	gometalinter.v2 --vendor --deadline=600s --enable-all --disable=lll ./...
 
 metalinter_test:
-	#protoc $(INCLUDE) --lint_out=. types/*.proto
```
Comment (melekes) [inline Makefile]: why remove proto linter???

File: types/types.proto [hunk 7]
```diff
@@ -3,25 +3,24 @@ package types;
...
-    RequestFlush flush = 2;
-    RequestInfo info = 3;
-    RequestSetOption set_option = 4;
-    RequestDeliverTx deliver_tx = 5;
-    RequestCheckTx check_tx = 6;
-    RequestCommit commit = 7;
-    RequestQuery query = 8;
-    RequestInitChain init_chain = 9;
-    RequestBeginBlock begin_block = 10;
+  oneof value {
```
Comment (melekes) [inline types/types.proto]: why this change?

File: Makefile (hunk 6 above)
Comment (adrianbrink) [reply Makefile]: Is proto linter fast enough to run on ci?

File: types/types.proto (hunk 7 above)
Comment (ethanfrey) [reply types/types.proto]: Just to screw with third party abci implementations?

File: Makefile (hunk 6 above)
Comment (jaekwon) [reply Makefile]: we don't need it, and it's a vulnerability to depend on it.

File: Makefile (hunk 5 above)
Comment (jaekwon) [reply Makefile]: lets rename metalinter_test to metalinter?

File: types/types.proto (hunk 7 above)
Comment (ebuchman) [reply types/types.proto]: Oh - probably to correspond with the numbering for Response

File: Makefile (hunk 6 above)
Comment (jaekwon) [reply Makefile]: like, all the stuff in GOTOOLS pretty much needs to be installed for anyone to start building, but that means installing something not used often... if that repo gets hacked, we've hacked all devs trying out ABCI.

Comment (jaekwon) [reply Makefile]: oh hmm... is --lint_out a default proto feature?  I removed a third party dependency from GOTOOLS and I thought --lint_out depended on that.

Comment (jaekwon) [reply Makefile]: Oops I forgot to remove https://github.com/ckaznocha/protoc-gen-lint as well.  That was the point.

File: types/types.proto (hunk 7 above)
Comment (jaekwon) [reply types/types.proto]: > Just to screw with third party abci implementations?

They should be using the protobuf file so hopefully they can just recompile.

Comment (jaekwon) [reply types/types.proto]: The order was changed to reflect the order of messages going into the app, more or less.

Comment (jaekwon) [reply types/types.proto]: after init_chain you can query, but doesn't make sense before then.  deliver is between begin and end block.  I think check_tx could go above begin_block, but it's nicer to keep those two side by side.  Commit comes at the very end.

File: Makefile (hunk 5 above)
Comment (melekes) [reply Makefile]: @zramsay made it a standard across all the repos. this might screw some things in the future.

Comment (jaekwon) [reply Makefile]: k thank you.  the order here was important because you need to protoc before you can compile/install, and it's nice to compile before you test.  (actually we should probably build before we test, and install after the tests pass, because install adds something to your GOPATH/bin. Then metalinter can come last because it's least important, and it gets in the way of development flow.

**Comment by melekes** [comment]
@ethanfrey want to do honor? just don't forget to pick `squash and merge` instead of default.

File: CHANGELOG.md [hunk 8]
```diff
@@ -1,5 +1,13 @@
 # Changelog
 
+## 0.9.0 (TBD)
+
+BREAKING CHANGES:
+ - [types] ResponseEndBlock: renamed Diffs field to ValidatorUpdates
```
Comment (ebuchman) [inline CHANGELOG.md]: also the field indices for the oneof Request and Response types

Comment (jaekwon) [reply CHANGELOG.md]: yes.

File: Makefile (hunk 5 above)
Comment (jaekwon) [reply Makefile]: that just seems like overkill.  can we just use one? https://github.com/golang/lint ?

Comment (jaekwon) [reply Makefile]: it's strange naming.  just call it `make metalinter` and `make metalinter_all`.
making it a goal to use all of metalinter's linters seems like overkill. I don't think that should be a goal.
